runtime/src/iree/hal/drivers/amdgpu/device/blit.c

```diff
@@ -0,0 +1,486 @@
+        &context->kernels->iree_hal_amdgpu_device_buffer_fill_block_x16;
+    element_size = IREE_HAL_AMDGPU_FILL_BLOCK_ELEMENT_SIZE;
+    block_size = IREE_HAL_AMDGPU_FILL_BLOCK_COUNT;
+  } else {
+    switch (pattern_length) {
+      case 1:
+        IREE_AMDGPU_TRACE_ZONE_APPEND_TEXT_LITERAL(z0, "fill_x1");
+        kernel_args = &context->kernels->iree_hal_amdgpu_device_buffer_fill_x1;
+        break;
+      case 2:
+        IREE_AMDGPU_TRACE_ZONE_APPEND_TEXT_LITERAL(z0, "fill_x2");
+        kernel_args = &context->kernels->iree_hal_amdgpu_device_buffer_fill_x2;
+        break;
+      case 4:
+        IREE_AMDGPU_TRACE_ZONE_APPEND_TEXT_LITERAL(z0, "fill_x4");
+        kernel_args = &context->kernels->iree_hal_amdgpu_device_buffer_fill_x4;
+        break;
+      case 8:
+        IREE_AMDGPU_TRACE_ZONE_APPEND_TEXT_LITERAL(z0, "fill_x8");
+        kernel_args = &context->kernels->iree_hal_amdgpu_device_buffer_fill_x8;
+        break;
+    }
+    element_size = pattern_length;
+    block_size = 1;
+  }
+
+  // Update kernargs (same API for all kernels).
+  const size_t element_count = length / element_size;
+  iree_hal_amdgpu_device_buffer_fill_kernargs_t* kernargs =
+      (iree_hal_amdgpu_device_buffer_fill_kernargs_t*)kernarg_ptr;
+  kernargs->target_ptr = target_ptr;
+  kernargs->element_length = element_count;
+  kernargs->pattern = pattern;
+
+  // To support fills with more than UINT_MAX elements (uint32_t grid_size)
+  // we divide the problem into chunks as needed. We keep the innermost chunk
+  // size small as if we do [X,Y,1] we're likely to overshoot and don't want to
+  // have too many wasted invocations.
+  const size_t block_count = IREE_AMDGPU_CEIL_DIV(element_count, block_size);
+  uint32_t grid_size_x = 1;
+  uint32_t grid_size_y = 1;
+  if (IREE_AMDGPU_LIKELY(block_count <= 0xFFFFFFFFu)) {
+    grid_size_x = (uint32_t)block_count;
+  } else {
+    grid_size_x = 256;
+    grid_size_y = (uint32_t)IREE_AMDGPU_CEIL_DIV(block_count, grid_size_x);
+  }
+
+  // Populate the packet.
+  const uint64_t queue_mask = context->queue->size - 1;  // power of two
+  iree_hsa_kernel_dispatch_packet_t* IREE_AMDGPU_RESTRICT dispatch_packet =
+      context->queue->base_address + (packet_id & queue_mask) * 64;
+  dispatch_packet->setup = kernel_args->setup;
+  dispatch_packet->workgroup_size[0] = kernel_args->workgroup_size[0];
+  dispatch_packet->workgroup_size[1] = kernel_args->workgroup_size[1];
+  dispatch_packet->workgroup_size[2] = kernel_args->workgroup_size[2];
+  dispatch_packet->reserved0 = 0;
+  dispatch_packet->grid_size[0] = grid_size_x;
+  dispatch_packet->grid_size[1] = grid_size_y;
+  dispatch_packet->grid_size[2] = 1;
+  dispatch_packet->private_segment_size = kernel_args->private_segment_size;
+  dispatch_packet->group_segment_size = kernel_args->group_segment_size;
+  dispatch_packet->kernel_object = kernel_args->kernel_object;
+  dispatch_packet->kernarg_address = kernarg_ptr;
+  dispatch_packet->reserved2 = 0;
```

Comment:
I see `reserved0` and `reserved2`, but no `reserved1`. Is that coming from the HSA spec? https://github.com/iree-org/iree/blob/7e273a1e2729f9c5588d290673dbcc8c3c4c0467/runtime/src/iree/hal/drivers/amdgpu/device/support/queue.h#L282-L305